ksqldb-engine/src/main/java/io/confluent/ksql/engine/TransientQueryCleanupService.java

```diff
@@ -0,0 +1,312 @@
+          Pattern.compile("(?i).*transient_.*_[0-9]\\d*_[0-9]\\d*");
+
+  private final BlockingQueue<Callable<Boolean>> stateDirsCleanupTasks;
+  private final BlockingQueue<Callable<Boolean>> topicsCleanupTasks;
+  private Set<String> localCommandsQueryAppIds;
+  private boolean isLocalCommandsInitialized;
+  private boolean isLocalCommandsProcessed;
+  private QueryRegistry queryRegistry;
+  private final String stateDir;
+  private final ServiceContext serviceContext;
+  private final int initialDelay;
+  private final int intervalPeriod;
+
+
+  public TransientQueryCleanupService(final ServiceContext serviceContext,
+                                      final KsqlConfig ksqlConfig) {
+    this.transientPattern = Pattern.compile(
+            ksqlConfig.getString(
+                    KsqlConfig.KSQL_TRANSIENT_QUERY_NAME_PREFIX_CONFIG));
+
+    this.initialDelay = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_INITIAL_DELAY_SECONDS);
+
+    this.intervalPeriod = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_PERIOD_SECONDS);
+
+    topicsCleanupTasks = new LinkedBlockingDeque<>();
+    stateDirsCleanupTasks = new LinkedBlockingDeque<>();
+
+    this.stateDir = ksqlConfig.getKsqlStreamConfigProps()
+            .getOrDefault(
+                    StreamsConfig.STATE_DIR_CONFIG,
+                    StreamsConfig.configDef()
+                            .defaultValues()
+                            .get(StreamsConfig.STATE_DIR_CONFIG))
+            .toString();
+
+    this.serviceContext = serviceContext;
+    this.isLocalCommandsInitialized = false;
+    this.isLocalCommandsProcessed = false;
+  }
+
+  @Override
+  protected void runOneIteration() {
+    if (isLocalCommandsInitialized && !isLocalCommandsProcessed) {
+      LOG.info("Adding LocalCommands to TransientQueryCleanupService.");
+      localCommandsQueryAppIds.forEach(id -> {
+        addTopicCleanupTask(new TransientQueryTopicCleanupTask(serviceContext, id));
+        addStateCleanupTask(new TransientQueryStateCleanupTask(id, stateDir));
+      });
+      isLocalCommandsProcessed = true;
+    }
+
+    LOG.info("Starting cleanup for leaked resources.");
+
+    topicsCleanupTasks.addAll(findPossiblyLeakedTransientTopics());
+    stateDirsCleanupTasks.addAll(findPossiblyLeakedStateDirs());
+
+    final int numLeakedTopics = topicsCleanupTasks.size();
+    final int numLeakedStateDirs = stateDirsCleanupTasks.size();
+
+    LOG.info("Cleaning up {} leaked topics.", numLeakedTopics);
+    cleanupLeakedTopics();
+
+    LOG.info("Cleaning up {} leaked state directories.", numLeakedStateDirs);
+    cleanupLeakedStateDirs();
+  }
+
+  @Override
+  public Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(initialDelay, intervalPeriod, TimeUnit.SECONDS);
+  }
+
+  public void cleanupLeakedTopics() {
+    try {
+      while (!topicsCleanupTasks.isEmpty()) {
```

Comment:
This won't exist the while loop until topicsCleanupTasks is empty, but what happens if you keep getting exceptions when running them? It looks like you're still enqueuing them again to immediately rerun.  That was my main comment last time.  I think you need two lists so that you let everything run once and anything that fails goes to the next round.

Reply:
updated to search for all leaked resources on every. If any deletion fasls then we just move on hoping that they would get discovered again in the next scheduled run